Replace remaining SupportPower CameraActors with CameraRange

reaperrr opened the pull request:
Using `RevealShroudEffect` like `NukePower` already does.

reaperrr committed:
Replace remaining SupportPower CameraActors with CameraRange

Using RevealShroudEffect like NukePower already does.

pchote reviewed (commented):
Could we define abstract `SupportPowerWithCamera`(+`Info`) superclasses to cut down on the code duplication?

MustaphaTR commented on mods/ra/rules/structures.yaml:
```diff
@@ -1352,7 +1352,6 @@ AFLD:
 		LongDesc: Reveals an area of the map\nand cloaked enemy units.
 		SelectTargetSpeechNotification: SelectTarget
 		EndChargeSpeechNotification: SpyPlaneReady
-		CameraActor: camera.spyplane
```
camera.spyplane has `DetectsCloaked`trait. Simply replacing that with CameraRange would drop it.

reaperrr replied:
Note to future reviewers: This is already fixed in the [3rd commit](https://github.com/OpenRA/OpenRA/pull/13858/commits/b9018cd7b884b7c46f51b2945067fda72f042de5) (though it's shown as 2nd commit by github).

reaperrr commented on mods/ra/rules/structures.yaml:
```diff
@@ -1352,7 +1352,6 @@ AFLD:
 		LongDesc: Reveals an area of the map\nand cloaked enemy units.
 		SelectTargetSpeechNotification: SelectTarget
 		EndChargeSpeechNotification: SpyPlaneReady
-		CameraActor: camera.spyplane
```
Hm, I see. Might need a DetectCloakedEffect.

pchote replied:
The simplest "fix" would be to change this to a paradrop power and drop a single camera.spyplane.

reaperrr commented:
> Could we define abstract `SupportPowerWithCamera`(+`Info`) superclasses to cut down on the code duplication?

I don't know how to make that inherit `SupportPower` while maintaining the abstract-ness, but how about we simply add the camera fields to `SupportPower[Info]`? The default camera range is zero anyway, and the implementation details are left to the non-abstract support powers as well, so that approach shouldn't have any disadvantages.

reaperrr committed:
Allow disabling ParaDrop.ChuteSound

reaperrr commented:
Updated.
I can't make heads or tails of the test/Travis errors, any idea @pchote?

pchote commented on mods/ra/rules/misc.yaml:
```diff
@@ -160,20 +160,33 @@ camera.paradrop:
+	Passenger:
+	Parachutable:
+		KilledOnImpassableTerrain: false
+		Image:
```
The linter is complaining because `GroundCorpseSequence` and `WaterCorpseSequence` refer to sequences on this image, which is being set to null.

pchote replied:
Take a look at the D2K paradrop power.  That drops units without having a visible parachute.

pchote commented:
Is help still wanted here?

reaperrr commented:
No, just forgot to remove the label.

reaperrr commented:
Rebased.

GraionDilach commented:
The mobile actors not occupying space thingie should be controllable somehow to fix that crate regression proper paradrop introduced.

reaperrr commented:
> The mobile actors not occupying space thingie should be controllable somehow to fix that crate regression proper paradrop introduced.

I don't know what regression you're referring to, and what do you mean with 'controllable' (since it's a boolean on Mobile already)?

abcdefg30 commented:
You cannot move a unit on the cell a crate is paradropped on anymore. You could do this before and collect the crate immediately.

GraionDilach commented:
Yes, that. I was in a hurry and had no time to look up the issue ticket - it was filed as #9555.

abcdefg30 commented:
Needs a rebase.

pchote commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -73,6 +73,9 @@ public class MobileInfo : ConditionalTraitInfo, IMoveInfo, IPositionableInfo, IF
 		public readonly bool SharesCell = false;
 
+		[Desc("Does this actor block other actors?")]
+		public readonly bool OccupiesSpace = true;
```
This really isn't very nice, IMO.  Can we please make the detector `Immobile` instead?

reaperrr replied:
Immobile doesn't implement IPositionable. IPositionable is required by Parachutable. Parachutable is required to move the detector to ground height.
And implementing IPositionable on Immobile was quite tricky last time I tried.

pchote commented:
Also still needs a rebase.

reaperrr commented:
Rebased and updated, but I get
> `System.InvalidOperationException`: TypeDictionary does not contain instance of type `OpenRA.LocationInit`

and need help debugging that.

reaperrr commented:
> [20:07] <+pchote> reaperrr: a simpler option would be to not create paradropped actors until they are dropped
[20:08] <+pchote> then we could use Immobile as-is

penev92 commented:
I don't like that last suggestion.
I can't debug, however, since the spawn positions are apparently broken, so I can't start a game properly.

reaperrr commented:
Updated.

reaperrr commented:
Rebased.

abcdefg30 commented:
Needs a rebase again.

reaperrr commented:
I guess there's no point in rebasing this until the upgrade rule refactor is merged.

reaperrr commented:
> I guess there's no point in rebasing this until the upgrade rule refactor is merged.

That didn't make it for the release and currently has no immediate ETA, so I rebased and removed the 'has dep' label.

Unit158 commented on OpenRA.Mods.Common/Traits/SupportPowers/ParatroopersPower.cs:
```diff
@@ -80,31 +73,25 @@ public Actor[] SendParatroopers(Actor self, WPos target, bool randomize = true,
 			if (!self.World.Map.Rules.Actors.TryGetValue(utLower, out unitType))
 				throw new YamlException("Actors ruleset does not include the entry '{0}'".F(utLower));
 
+			// Cache target before aircraft altitude is added to spawn camera on ground
```
Im confused by this, I thought the camera was removed o.O

reaperrr replied:
Unlucky wording, 'camera' should really be 'camera effect' or 'camera reveal'.
Basically, without this the shroud would be revealed around the aircrafts screen position (in other words, a few cells north of the target cell).

reaperrr replied:
Adding just in case: only the camera *actor* spawning was removed, all support powers now support revealing via effect instead.

abcdefg30 commented on OpenRA.Mods.Common/Traits/SupportPowers/ParatroopersPower.cs:
```diff
@@ -126,17 +113,9 @@ public Actor[] SendParatroopers(Actor self, WPos target, bool randomize = true,
 				{
-					if (camera != null)
-					{
-						camera.QueueActivity(new Wait(info.CameraRemoveDelay));
```
Before this change the camera was only removed after the last plane had left the area. We have to update the remove delay to compensate for that.

reaperrr committed:
Factor in Cordon and SquadSize into air power camera remove delays

This estimates the time between first aircraft entering and last aircraft leaving the target cordon, to make sure the camera reveal isn't removed before the last plane exits the cordon.

reaperrr commented:
@abcdefg30 Updated.

reaperrr commented:
Rebased and removed spyplane stealth detection work-around as per IRC comments.

In my opinion it would be better to implement proper stealth detection without using actors anyway, but that can wait for a later PR.

reaperrr commented:
Since this PR currently has no significant practical benefit and actually reduces flexibility a bit, I'm closing this until we either decide we need this for some other reason, or find some solution to avoid the flexibility loss.
